src/main/java/org/elasticsearch/search/aggregations/InternalMultiBucketAggregation.java

```diff
@@ -58,20 +81,30 @@ public Object getProperty(String containingAggName, List<String> path) {
             String aggName = path.get(0);
             if (aggName.equals("_count")) {
                 if (path.size() > 1) {
-                    throw new ElasticsearchIllegalArgumentException("_count must be the last element in the path");
+                    throw new InvalidAggregationPathException("_count must be the last element in the path");
                 }
                 return getDocCount();
             } else if (aggName.equals("_key")) {
                 if (path.size() > 1) {
-                    throw new ElasticsearchIllegalArgumentException("_key must be the last element in the path");
+                    throw new InvalidAggregationPathException("_key must be the last element in the path");
                 }
                 return getKey();
             }
             InternalAggregation aggregation = aggregations.get(aggName);
             if (aggregation == null) {
-                throw new ElasticsearchIllegalArgumentException("Cannot find an aggregation named [" + aggName + "] in [" + containingAggName + "]");
+                throw new InvalidAggregationPathException("Cannot find an aggregation named [" + aggName + "] in [" + containingAggName
+                        + "]");
             }
             return aggregation.getProperty(path.subList(1, path.size()));
         }
     }
+
+    public static abstract class Factory<A extends InternalMultiBucketAggregation, B extends InternalMultiBucketAggregation.InternalBucket> {
+
+        public abstract String type();
+
+        public abstract A create(List<B> buckets, A prototype);
```

Comment:
Yep, in fact the whole class is unnecessary since we moved the create and createBucket methods into the Aggregations themselves rather than having separate factory classes